### ISSUE TITLE
non-defined -x tests behaviour unexpected ( "-1" eq -F 1' )

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6842]
Migrated from [rt.perl.org#24212](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=24212) (status was 'resolved')

Searchable as RT24212$

**Comment by p5pRT** [comment]
### From david.dyck@fluke.com

#### Created by david.dyck@fluke.com



I had a error in some code the other day\, when
I intended to code \-f but instead I entered \-F\.

To my surprise\, I didn't get an error or warning\.

After reading toke\.c in the perl sources\, I see 
the following comment when the letter after the '\-' in \-x
is not one those listed in perldoc \-f \-x

&nbsp;    /\* Assume it was a minus followed by a one\-letter named
&nbsp;     \* subroutine call \(or a \-bareword\)\, then\. \*/

In perlop\.pod


&nbsp;    Unary "\-" performs arithmetic negation if the operand is numeric\. If the
&nbsp;    operand is an identifier\, a string consisting of a minus sign concatenated
&nbsp;    with the identifier is returned\. Otherwise\, if the string starts with a
&nbsp;    plus or minus\, a string starting with the opposite sign is returned\. One
&nbsp;    effect of these rules is that "\-bareword" is equivalent to "\-bareword"\.

&nbsp;  \[ perhaps the text above could be written differently\, as
&nbsp;    "\-bareword" is alwasy equal to "\-bareword"\,  \(I do see that
&nbsp;    the pod source encodes this differently\, as
&nbsp;      C\<\-bareword> is equivalent to C\<"\-bareword">
&nbsp;    which in html displays as
&nbsp;      \-bareword is equivalent to "\-bareword"\. \]

Anyway\, I expect that perl should give me an error\, a warning\,
or return a bareword converted string\, but in the following
examples perl is definitely DWIM'ing

If I create a one\-letter subroutine I see that the sub is not called&#8203;:

$ perl \-wle 'use strict; print \-F 1'
\-1

And the \-F is not converted to a \-bareword

$ perl \-wle 'use strict; print \-F 1'
\-1

$ perl \-wle 'use strict; print "strange" if "\-1" eq \-F 1'
strange


The follow scripts reports the set of letters where \-x doesn't
do what I expect\.
&nbsp;  e\.g\. strange \-X codes&#8203;: DEFGHIJKLNOPQUVYZahijmnoqvy


\#\!/usr/local/bin/perl \-w

use strict;
no strict 'refs';
use B&#8203;::Deparse;
my $d=B&#8203;::Deparse\->new\(\);
my $q = "1"; 
my @&#8203;testlist = \("a" \.\."z"\, "A" \.\. "Z"\);
my %strange;
for my $x \(@&#8203;testlist\) \{
&nbsp;    my $s =  "\{\\n    \-$x \\$q;\\n\}";
&nbsp;    my $sr = eval "sub $s"; 
&nbsp;    warn "$@&#8203;" if length $@&#8203;; 

&nbsp;    my $dp = $d\->coderef2text\( $sr \); 
&nbsp;    if \($s ne $dp\) \{ 
&nbsp;    print "expect&#8203;:\\n$s\\ndeparse returned&#8203;:\\n$dp\\n"; 
&nbsp;    \+\+$strange\{$x\}
&nbsp;    \} 
\} 
print "strange \-X codes&#8203;: "\,join\( ""\, sort keys %strange\)\, "\\n";
\_\_END\_\_

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.9.0:

Configured by dcd at Tue Oct  7 12:13:57 PDT 2003.

Summary of my perl5 (revision 5.0 version 9 subversion 0 patch 21419) configuration:
  Platform:
    osname=linux, osvers=2.4.23pre1, archname=i686-linux
    uname='linux dd 2.4.23pre1 #2 wed aug 27 11:20:37 pdt 2003 i686 '
    config_args='-Dmksymlinks -Dinstallusrbinperl -Uversiononly -Dusedevel -Doptimize=-O3 -g -de -Dcf_email=david.dyck@fluke.com'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3 -g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='egcs-2.91.66.1 19990314/Linux (egcs-1.1.2 release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -ldbm -ldb -ldl -lm -lc
    perllibs=-ldl -lm -lc
    libc=/lib/libc.so.5.4.44, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL21398


@INC for perl v5.9.0:
    /usr/local/lib/perl5/5.9.0/i686-linux
    /usr/local/lib/perl5/5.9.0
    /usr/local/lib/perl5/site_perl/5.9.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.9.0
    /usr/local/lib/perl5/site_perl/5.8.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.9.0:
    HOME=/home/dcd
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/dcd/bin:/sbin:/usr/local/bin:/bin:/usr/bin:/usr/X11/bin:/usr/games:/usr/local/samba:/home/hobbes/tools/scripts:/home/hobbes/tools/linux:/usr0/hobbes/tools/scripts:/usr0/dcd/bin:/apps/general/bin:/usr/public
    PERL5_CPANPLUS_CONFIG=/home/dcd/.cpanplus/config
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

David Dyck \(via RT\) \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> 
> $ perl \-wle 'use strict; print \-F 1'
> \-1
> 
> $ perl \-wle 'use strict; print "strange" if "\-1" eq \-F 1'
> strange

Thanks for the report\, fixed in bleadperl as change \#21452\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
